### PR TITLE
Read EEPROM from file

### PR DESCRIPTION
**Pull request by dracc**
Touches #28, but does not quite go the full distance.
Code is more or less @mborgerson's bootrom work, just refitted for this purpose.

What works:
- [x] Load EEPROM from file
- [x] Use bunnies EEPROM if file not supplied

### CODE REVIEW

File: hw/xbox/xbox.c [hunk 1]
```diff
@@ -253,7 +253,35 @@ void nv2a_init(PCIBus *bus, int devfn, MemoryRegion *ram);
...
+
+    const char *eeprom_file = object_property_get_str(qdev_get_machine(),
+                                                      "eeprom", NULL);
+    if ((eeprom_file != NULL) && *eeprom_file) {
+        filename = qemu_find_file(QEMU_FILE_TYPE_BIOS, eeprom_file);
+        assert(filename);
+
+        eeprom_size = get_image_size(filename);
+        assert(eeprom_size <= 256);
+        if (eeprom_size < 256) {
```
Comment (mborgerson) [inline hw/xbox/xbox.c]: EEPROM should always be 256 bytes.

File: hw/xbox/xbox.c [hunk 2]
```diff
@@ -253,7 +253,35 @@ void nv2a_init(PCIBus *bus, int devfn, MemoryRegion *ram);
 /* PC hardware initialisation */
 static void xbox_init(MachineState *machine)
 {
-    xbox_init_common(machine, default_eeprom, NULL, NULL);
+    char *filename;
```
Comment (mborgerson) [inline hw/xbox/xbox.c]: This works for Xbox machine but would have to be duplicated for Chihiro support. We should think of a common solution so both can support it via the same code.

File: hw/xbox/xbox.c [hunk 3]
```diff
@@ -253,7 +253,35 @@ void nv2a_init(PCIBus *bus, int devfn, MemoryRegion *ram);
...
+    if ((eeprom_file != NULL) && *eeprom_file) {
+        filename = qemu_find_file(QEMU_FILE_TYPE_BIOS, eeprom_file);
+        assert(filename);
+
+        eeprom_size = get_image_size(filename);
+        assert(eeprom_size == 256);
+
+        eeprom_data = g_malloc(eeprom_size);
+        fd = open(filename, O_BINARY);
+        assert(fd >= 0);
```
Comment (mborgerson) [inline hw/xbox/xbox.c]: I know the bootrom code does this exact same thing, but I think (in both cases) we should fail more gracefully here and print an error message saying that the file could not be opened. It's very possible that we don't have permission or something.

File: hw/xbox/xbox.c [hunk 4]
```diff
@@ -253,7 +253,35 @@ void nv2a_init(PCIBus *bus, int devfn, MemoryRegion *ram);
...
+    uint8_t *eeprom_data;
+
+    const char *eeprom_file = object_property_get_str(qdev_get_machine(),
+                                                      "eeprom", NULL);
+    if ((eeprom_file != NULL) && *eeprom_file) {
+        filename = qemu_find_file(QEMU_FILE_TYPE_BIOS, eeprom_file);
+        assert(filename);
+
+        eeprom_size = get_image_size(filename);
+        assert(eeprom_size == 256);
```
Comment (mborgerson) [inline hw/xbox/xbox.c]: Same as below, we should fail more gracefully. It's highly likely the user will pass some invalid file here and we will just crash.

File: hw/xbox/xbox.c [hunk 5]
```diff
@@ -253,7 +253,35 @@ void nv2a_init(PCIBus *bus, int devfn, MemoryRegion *ram);
 /* PC hardware initialisation */
 static void xbox_init(MachineState *machine)
 {
-    xbox_init_common(machine, default_eeprom, NULL, NULL);
+    char *filename;
+    int fd;
+    int eeprom_size = 256;
```
Comment (mborgerson) [inline hw/xbox/xbox.c]: Make this a const and compare against it below.

File: hw/xbox/xbox.c [hunk 6]
```diff
@@ -253,7 +253,53 @@ void nv2a_init(PCIBus *bus, int devfn, MemoryRegion *ram);
...
+            g_free(filename);
+            exit(1);
+            return;
+        }
+
+        close(fd);
+        g_free(filename);
+    } else {
+        memcpy(eeprom_data, default_eeprom, eeprom_size);
+    }
```
Comment (JayFoxRox) [inline hw/xbox/xbox.c]: I think the complexity of this code section warrants putting it in a separate function (~`load_eeprom` or something)

File: hw/xbox/xbox.c [hunk 7]
```diff
@@ -253,7 +253,53 @@ void nv2a_init(PCIBus *bus, int devfn, MemoryRegion *ram);
...
+    char *filename;
+    int fd;
+    int rc;
+    int eeprom_file_size;
+    const int eeprom_size = 256;
+    uint8_t *eeprom_data = g_malloc(eeprom_size);
+
+    const char *eeprom_file = object_property_get_str(qdev_get_machine(),
+                                                      "eeprom", NULL);
+    if ((eeprom_file != NULL) && *eeprom_file) {
```
Comment (JayFoxRox) [inline hw/xbox/xbox.c]: nit: is `*eeprom_file` really necessary and the QEMU way of doing things?

File: hw/xbox/xbox.c [hunk 8]
```diff
@@ -253,7 +253,53 @@ void nv2a_init(PCIBus *bus, int devfn, MemoryRegion *ram);
...
+        filename = qemu_find_file(QEMU_FILE_TYPE_BIOS, eeprom_file);
+        assert(filename);
+
+        eeprom_file_size = get_image_size(filename);
+        if (eeprom_size != eeprom_file_size) {
+            fprintf(stderr,
+                    "qemu: EEPROM file size != 256 bytes. (Is %i bytes)\n",
```
Comment (JayFoxRox) [inline hw/xbox/xbox.c]: Use `"...%d...", ..., eeprom_size)`

File: hw/xbox/xbox.c (hunk 6 above)
Comment (mborgerson) [reply hw/xbox/xbox.c]: Agree (what I was hinting at with above comment too)

File: hw/xbox/xbox.c (hunk 7 above)
Comment (mborgerson) [reply hw/xbox/xbox.c]: Yes, it is necessary. What are you suggesting?

File: hw/xbox/xbox.c [hunk 9]
```diff
@@ -253,7 +253,53 @@ void nv2a_init(PCIBus *bus, int devfn, MemoryRegion *ram);
...
+                    eeprom_file_size);
+            g_free(filename);
+            exit(1);
+            return;
+        }
+
+        fd = open(filename, O_BINARY);
```
Comment (JayFoxRox) [inline hw/xbox/xbox.c]: nit: Open as read-only?

**Comment by JayFoxRox** [comment]
- Commits should be renamed to prefix style ("eeprom: ..." or something)
- EEPROM image should be provided somewhere (this should happen before this is merged, so we get closer to getting rid of the built-in image)

I think write can happen in a follow-up PR.
I always considered this PR WIP because it shows 2/3 tasks complete, but it looks like you are not actively working on this anymore.

File: hw/xbox/xbox.c (hunk 7 above)
Comment (JayFoxRox) [reply hw/xbox/xbox.c]: `strlen(eeprom_file) > 0` or something?

I think it's just not very self-documenting.

Comment (mborgerson) [reply hw/xbox/xbox.c]: Ah yes, agree

File: hw/xbox/xbox.c [hunk 10]
```diff
@@ -150,7 +150,7 @@ static void xbox_flash_init(MemoryRegion *rom_memory)
     const char *bootrom_file = object_property_get_str(qdev_get_machine(),
                                                        "bootrom", NULL);
 
-    if ((bootrom_file != NULL) && *bootrom_file) {
+    if ((bootrom_file != NULL) && *botrom_file) {
```
Comment (mborgerson) [inline hw/xbox/xbox.c]: ?

Comment (dracc) [reply hw/xbox/xbox.c]: I made the edit from my cellphone, this change was unintended. Will fix asap

File: hw/xbox/xbox.c [hunk 11]
```diff
@@ -227,11 +227,62 @@ static void xbox_memory_init(PCMachineState *pcms,
...
+    int fd;
+    int rc;
+    int eeprom_file_size;
+    const int eeprom_size = 256;
+
+    uint8_t *eeprom_data = g_malloc(eeprom_size);
+
+    const char *eeprom_file = object_property_get_str(qdev_get_machine(),
+                                                      "eeprom", NULL);
+    if ((eeprom_file != NULL) && (strlen(eeprom_file) > 0)){
```
Comment (JayFoxRox) [inline hw/xbox/xbox.c]: Please undo this change - the original code was more similar to existing code (the `bootrom_file`).
If `*eeprom_file` is the standard way of doing it, we should keep doing so (or change *all* instances at once - which doesn't fit into this PR).

File: hw/xbox/xbox.c [hunk 12]
```diff
@@ -227,11 +227,62 @@ static void xbox_memory_init(PCMachineState *pcms,
     xbox_flash_init(rom_memory);
 }
 
+uint8_t* load_eeprom(void)
```
Comment (mborgerson) [inline hw/xbox/xbox.c]: Nit: `*` goes to the immediate left of the identifier (e.g. `uint8_t *load_eeprom`). Repeated a couple times below.

File: hw/xbox/xbox.c [hunk 13]
```diff
@@ -227,11 +227,62 @@ static void xbox_memory_init(PCMachineState *pcms,
...
+        if (eeprom_size != eeprom_file_size) {
+            fprintf(stderr,
+                    "qemu: EEPROM file size != %d bytes. (Is %d bytes)\n",
+                    eeprom_size, eeprom_file_size);
+            g_free(filename);
+            exit(1);
+            return NULL;
+        }
+
+        fd = open(filename, O_RDONLY);
```
Comment (JayFoxRox) [inline hw/xbox/xbox.c]: nit: might still want the binary flag to not break on shitty platforms (Windows in particular). not sure which `open` variant is exposed here and if QEMU deals with this.

**Comment by dracc** [comment]
Review comments taken care of.
Regarding EEPROM binary distribution; could bunnies EEPROM be distributed in the same vein as the HDD image?

File: hw/xbox/xbox.c [hunk 14]
```diff
@@ -227,11 +227,62 @@ static void xbox_memory_init(PCMachineState *pcms,
...
+                    eeprom_size, eeprom_file_size);
+            g_free(filename);
+            exit(1);
+            return NULL;
+        }
+
+        fd = open(filename, O_RDONLY | O_BINARY);
+        if (fd < 0) {
+            fprintf(stderr, "qemu: EEPROM file '%s' could not be opened.\n", filename);
+            g_free(filename);
```
Comment (mborgerson) [inline hw/xbox/xbox.c]: Not `free`ing `eeprom_data`, but it doesn't really matter here because we are exiting.

**Comment by mborgerson** [comment]
Thanks @dracc

File: hw/xbox/xbox.c (hunk 14 above)
Comment (dracc) [reply hw/xbox/xbox.c]: Should still be considered a leak, no?

Comment (JayFoxRox) [reply hw/xbox/xbox.c]: Yes, and some cleanup should fix it eventually because it just polutes valgrind logs.
